### ISSUE TITLE
Widevine license request fails with HTTP 422 on a Xiaomi Redmi 9

### ISSUE THREAD

**Issue by Stonos** [google/ExoPlayer#8373]
Our app plays Widevine-encrypted DASH streams using ExoPlayer 2.12.1. We have a few users that have reported playback issues, and most of them are using Xiaomi devices.

We managed to get a bugreport from a user that is experiencing issues with a Xiaomi Redmi 9 (galahad) running Android 10. 
Here is the exception that occurs (I have sent the full bugreport via email):
```
internalError [eventTime=12.96, mediaPos=3581.00, window=0, period=0, drmSessionManagerError
  com.google.android.exoplayer2.drm.MediaDrmCallbackException: com.google.android.exoplayer2.upstream.HttpDataSource$InvalidResponseCodeException: Response code: 422
      at q2.k.a.b.s2.m0.b(SourceFile:30)
      at q2.k.a.b.s2.m0.a(SourceFile:15)
      at q2.k.a.b.s2.p.handleMessage(SourceFile:3)
      at android.os.Handler.dispatchMessage(Handler.java:107)
      at android.os.Looper.loop(Looper.java:224)
      at android.os.HandlerThread.run(HandlerThread.java:67)
  Caused by: com.google.android.exoplayer2.upstream.HttpDataSource$InvalidResponseCodeException: Response code: 422
      at q2.k.a.b.t2.b.a.a(SourceFile:53)
      at q2.k.a.b.d3.k0.a(SourceFile:3)
      at q2.k.a.b.d3.o.read(SourceFile:6)
      at q2.k.a.b.d3.o.read(SourceFile:3)
      at q2.k.a.b.e3.j0.Z(SourceFile:2)
      at q2.k.a.b.s2.m0.b(SourceFile:8)
      ... 5 more
]
```

It looks like the DRM server responds with HTTP 422 for some reason. We asked our DRM vendor about this, and this is what they said:

> Generally HTTP 422 is a standard error thrown when the content-type parameter is not correctly setup to match the actual content that is being sent. It's usually an autogenerated error when this conflict is detected.
> 
> On our services HTTP 422 error is also thrown when the payload cannot be processed due to several reasons (depending on the data service). For example the payload can be syntactically correct but for example what you are trying to do is already done e.g. publishing

What could be causing this issue? Other devices work fine.
Thanks.

**Comment by ojw28** [comment]
@jt64 - Are you able to shed any light on this? Is the `content-type` parameter mentioned by the DRM vendor part of the key request, and would this be indicative of an incorrect CDM implementation?

**Comment by rrfrias** [comment]
The content-type is provided either by the app or Exoplayer. There are no errors in the bugreport related to this failures and there is not much more we can tell from the logs. Can you work with the license service to identify why the request is being rejected with a HTTP 422 error?

**Comment by BucherTomas** [comment]
Not sure if we are seeing the same issue, but we also have a couple of Xiaomi Redmi 9 devices (also their Note variant) that are unable to decrypt and play media, while others from the same device group work fine.

We asked our license provider and they said that the license request is using invalid signature, so possibly it is the same reason for OP's DRM provider. I sent the payload to dev.exoplayer@gmail.com.

This appears to be a specific device CDM issue, but when we asked one user, they confirmed that they can play the same content in mobile browser instead of our ExoPlayer-based application using the same license details. User's DRM info screenshot is attached, they are running "merlinnfc" instead of OP's "galahad", but I assume the codebase would be more or less the same.

![Redmi](https://user-images.githubusercontent.com/18445121/104894295-c8826800-5974-11eb-9ccf-10fee48fa6f6.jpg)
![Redmi2](https://user-images.githubusercontent.com/18445121/104894301-cddfb280-5974-11eb-8fcb-6bd574a4fe25.jpg)


Another affected user we have some data from has the same Android version QP1A.190711.020 and the same security patch from Nov 1st, 2020, so there might be some connection. We cannot say, if the other working Redmi 9 devices have the same build fingerprint, we just see that they are also running Android 10.

Do you think you would be able to consult this with Xiaomi devs? Is Widevine L3 type request signed differently than L1 request? Is that why playback in mobile browser worked, because maybe it fell back to L3 for the browser and it helped?

Thank you.

**Comment by ojw28** [comment]
@rrfrias - Is the payload that @BucherTomas sent to `dev.exoplayer@gmail.com` useful for root causing this?

**Comment by rrfrias** [comment]
> We asked our license provider and they said that the license request is using invalid signature

Thanks for this information. We are tracking this issue (internal ref: 151266089). The license request is incorrectly signed by the TEE (Trusted Execution Environment). We have had reports of this on other Xiaomi devices as well (Redmi Note 8 and Note 8 pro).  

We are working with Xiaomi, but one part that has been hindering progress is being able to reproduce this internally. We understand that request is not signed correctly but Xiaomi cannot tell why. 

I received a device from a partner that exhibited this issue last week. They signed out of the device. Now that i sign in, i am unable to reproduce. That indicates that it might be tied to the provisioning process rather than the device and that a reprovision might clear the issue. We are investigating this and a few other options.

**Comment by Stonos** [comment]
Unfortunately the DRM provider wasn't able to provide us with any additional information.

> We have had reports of this on other Xiaomi devices as well (Redmi Note 8 and Note 8 pro).

We've also had reports with some Redmi Note 8 Pros not working, but we didn't manage to receive a bugreport to confirm if it's the same issue. Sounds like it is.

> I received a device from a partner that exhibited this issue last week. They signed out of the device. Now that i sign in, i am unable to reproduce. That indicates that it might be tied to the provisioning process rather than the device and that a reprovision might clear the issue. We are investigating this and a few other options.

This is interesting! Does provisioning take place after you login with a Google account? Could affected users solve this issue by logging out and logging in again with the same Google account?

**Comment by rrfrias** [comment]
> Could affected users solve this issue by logging out and logging in again with the same Google account?

Probably not. But a factory reset might work. We are trying to figure out a less intrusive alternative.

**Comment by BucherTomas** [comment]
@rrfrias Is there any progress in your ideas how to solve this situation in the future? Thank you.

**Comment by rrfrias** [comment]
We have made some progress but we do not have an update at this time. If the user does not have offline licenses, reprovisioning might help. There is [no API for that yet](https://github.com/google/ExoPlayer/issues/8039). Will update when we something actionable.

**Comment by rrfrias** [comment]
Xiaomi has issued software updates for the Redmi Note family of devices. 

The Redmi Note 9 
* android 10: should have addressed this issue with a December 2020 update
* android 11: should have contained the fix from launch.

Devices that previously encountered this error when requesting a license will continue to fail until provisioning is attempted again. One way is to force provisioning again by throwing NotProvisionedException from inside MediaDrmCallback.executeKeyRequest.

This should also help devices without the software update. For devices that have this problem these failures are seen in about 0.4% to 2.5% of provisioning attempts. Let us know if you still see problems after this.

**Comment by BucherTomas** [comment]
@rrfrias thank you very much for the update. Did Xiaomi or can you, however, provide any hints about how to identify these circumstances so that the app can throw NotProvisionedException?

Because the device presents itself as provisioned, the issue manifests itself only after a license is acquired and at that moment it is already late. The DRM subsystem then does not propagate the actual reason to the app, I would assume that it appears only as fairly generic DRM exception that the license is denied or simply not acquired, but that can be caused by various reasons.

**Comment by rrfrias** [comment]
The bug causes an invalid signature in the license request. The invalid signature is detected at the license service and a 106 error is returned. It is up to the partners license service to return an error. In this case the license service returns an HTTP 422 error. If the error returned to the app is unique enough that it is able to identify a invalid signed license request a re-provisioning can be forced.

We considering including an error in the license response so that the partner's license service/app does not have to handle it out of band. For now the app might need to handle it as mentioned above.

**Comment by rrfrias** [comment]
If you are a Certificated Widevine Integration Partner you should have access to developer documentation that details how to handle an Invalid Signature error (106) at the License server (see a section in Widevine custom extensions to MediaDrm). This may occur when a license request is processed. Instead of sending back an HTTP error, the license service should send a message to the app to indicate that reprovisioning is needed. Here is a code snippet to force a provisioning.

```
HttpMediaDrmCallback drmCallback =
  new HttpMediaDrmCallback(/* defaultLicenseUrl= */ null, new DefaultHttpDataSource.Factory());

try (MediaDrm mediaDrm = new MediaDrm(C.WIDEVINE_UUID)) {
  ProvisionRequest provisionRequest = mediaDrm.getProvisionRequest();
  mediaDrm.provideProvisionResponse(
      drmCallback.executeProvisionRequest(
      C.WIDEVINE_UUID,
      new ExoMediaDrm.ProvisionRequest(
          provisionRequest.getData(), provisionRequest.getDefaultUrl())));
}
```

Let me know if you have any more questions otherwise I will close this in 2 weeks.
